src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ParameterCapturingService.cs

```diff
@@ -1,84 +1,316 @@
+
+            public void Dispose()
+            {
+                ProbeManager.Dispose();
+            }
+        }
+
         private long _disposedState;
-        private readonly bool _isAvailable;
 
-        private readonly FunctionProbesManager? _probeManager;
-        private readonly ILogger? _logger;
+        private ParameterCapturingEvents.ServiceState _serviceState;
+        private string _serviceStateDetails = string.Empty;
+
+        private readonly InitializedState? _initializedState;
+
+        private readonly ParameterCapturingEventSource _eventSource = new();
 
         public ParameterCapturingService(IServiceProvider services)
         {
-            _logger = services.GetService<ILogger<ParameterCapturingService>>();
-            if (_logger == null)
+            try
+            {
+                ArgumentNullException.ThrowIfNull(SharedInternals.MessageDispatcher);
+
+                // Register the command callbacks (if possible) first so that dotnet-monitor
+                // can be notified of any initialization errors when it tries to invoke the commands.
+                SharedInternals.MessageDispatcher.RegisterCallback<StartCapturingParametersPayload>(
+                    IpcCommand.StartCapturingParameters,
+                    OnStartMessage);
+
+                SharedInternals.MessageDispatcher.RegisterCallback<StopCapturingParametersPayload>(
+                    IpcCommand.StopCapturingParameters,
+                    OnStopMessage);
+
+                _initializedState = new InitializedState(services);
+            }
+            catch (NotSupportedException ex)
+            {
+                ChangeServiceState(ParameterCapturingEvents.ServiceState.NotSupported, ex.Message);
+            }
+            catch (Exception ex)
+            {
+                UnrecoverableError(ex);
+            }
+        }
+
+        private bool IsAvailable()
+        {
+            return _serviceState == ParameterCapturingEvents.ServiceState.Running;
+        }
+
+        private void OnStartMessage(StartCapturingParametersPayload payload)
+        {
+            if (!IsAvailable() || _initializedState == null)
+            {
+                BroadcastServiceState();
+                return;
+            }
+
+            if (payload.Methods.Length == 0)
+            {
+                _eventSource.FailedToCapture(payload.RequestId, new ArgumentException(nameof(payload.Methods)));
+                return;
+            }
+
+            CapturingRequest request = new(payload);
+            if (!_initializedState.AllRequests.TryAdd(payload.RequestId, request))
+            {
+                _eventSource.FailedToCapture(payload.RequestId, new ArgumentException(nameof(payload.RequestId)));
+                return;
+            }
+
+            if (_initializedState.RequestQueue.Writer.TryWrite(request) != true)
+            {
+                _initializedState.AllRequests.TryRemove(payload.RequestId, out _);
```

Comment:
I think that `request.StopRequest.TrySetCanceled` should also be called to make sure nothing is actually waiting on the abandoned request.